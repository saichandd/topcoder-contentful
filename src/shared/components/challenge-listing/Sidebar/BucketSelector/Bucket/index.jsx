/**
 * Regular sidebar row.
 */

import _ from 'lodash';
import { challenge as challengeUtils } from 'topcoder-react-lib';
import PT from 'prop-types';
import React from 'react';
import './style.scss';

const Filter = challengeUtils.filter;

export default function Bucket({
  active,
  bucket,
  challenges,
  disabled,
  onClick,
}) {
  let count;
  if (!bucket.hideCount && !disabled) {
    const filter = Filter.getFilterFunction(bucket.filter);
    count = challenges.filter(filter).length;
    count = (
      <span styleName="right">
        {count}
      </span>
    );
  }

  const error = Boolean(bucket.error) && (
    <div styleName="errorMsg">
      {bucket.error}
    </div>
  );

  if (active) {
    return (
      <div styleName="active bucket">
        {bucket.name}
        {count}
        {error}
      </div>
    );
  }

  return (
    <div
      onClick={disabled ? _.noop : onClick}
      onKeyPress={e => (e.key === 'Enter' ? onClick() : null)}
      role="button"
      styleName="bucket"
      tabIndex={0}
    >
      {bucket.name}
      {count}
      {error}
    </div>
  );
}

Bucket.defaultProps = {
  active: false,
  disabled: false,
  onClick: _.noop,
};

Bucket.propTypes = {
  active: PT.bool,
  bucket: PT.shape({
    hideCount: PT.bool,
    name: PT.string.isRequired,
    error: PT.string,
  }).isRequired,
  challenges: PT.arrayOf(PT.shape).isRequired,
  disabled: PT.bool,
  onClick: PT.func,
};
