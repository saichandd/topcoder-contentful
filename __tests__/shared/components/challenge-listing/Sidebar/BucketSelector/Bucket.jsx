import React from 'react';
import Renderer from 'react-test-renderer/shallow';
import TU from 'react-dom/test-utils';
import Bucket from 'components/challenge-listing/Sidebar/BucketSelector/Bucket';

test('Matches shallow shapshot', () => {
  const renderer = new Renderer();
  renderer.render((
    <Bucket challenges={[]} disabled bucket={{ name: 'name' }} />
  ));
  expect(renderer.getRenderOutput()).toMatchSnapshot();
  renderer.render((
    <Bucket challenges={[]} disabled bucket={{ name: 'name' }} active />
  ));
  expect(renderer.getRenderOutput()).toMatchSnapshot();
});


class Wrapper extends React.Component {
  componentDidMount() {}

  render() {
    return <Bucket {...this.props} />;
  }
}

const onClick = jest.fn();

test('handle key press', () => {
  const instance = TU.renderIntoDocument((<Wrapper challenges={[]} onClick={onClick} bucket={{ name: 'name' }} />));
  const matches = TU.findAllInRenderedTree(instance, item => item && item.className && item.className.match('bucket'));
  expect(matches).toHaveLength(1);
  expect(onClick).toHaveBeenCalledTimes(0);
  TU.Simulate.keyPress(matches[0], { key: 'Enter' });
  expect(onClick).toHaveBeenCalledTimes(1);
  TU.Simulate.keyPress(matches[0], { key: 'A' });
  expect(onClick).toHaveBeenCalledTimes(1);
});
