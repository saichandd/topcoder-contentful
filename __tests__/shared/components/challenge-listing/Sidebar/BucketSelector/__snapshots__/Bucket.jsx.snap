// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches shallow shapshot 1`] = `
<div
  className="src-shared-components-challenge-listing-Sidebar-BucketSelector-Bucket-___style__bucket___1M3_m"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="button"
  tabIndex={0}
>
  name
</div>
`;

exports[`Matches shallow shapshot 2`] = `
<div
  className="src-shared-components-challenge-listing-Sidebar-BucketSelector-Bucket-___style__active___3-5kD src-shared-components-challenge-listing-Sidebar-BucketSelector-Bucket-___style__bucket___1M3_m"
>
  name
</div>
`;
